sklearn/calibration.py

```diff
@@ -311,145 +326,150 @@ def _more_tags(self):
 
-        sample_weight : array-like of shape (n_samples,), default=None
-            Sample weights. If None, then samples are equally weighted.
+    X : array-like
+        Sample data used to calibrate predictions.
 
-        Returns
-        -------
-        self : object
-            Returns an instance of self.
-        """
+    y : ndarray, shape (n_samples,)
+        The targets.
+
+    sample_weight : ndarray, shape (n_samples,), default=None
+        Sample weights. If `None`, then samples are equally weighted.
 
-        self.label_encoder_ = LabelEncoder()
-        if self.classes is None:
-            self.label_encoder_.fit(y)
+    Returns
+    -------
+    pipeline : _CalibratedClassiferPipeline instance
+    """
+    Y = label_binarize(y, classes=label_encoder_.classes_)
+    df, pos_class_indices = _get_predictions(clf_fitted, X, label_encoder_)
+
+    calibrated_classifiers = []
+    for class_idx, this_df in zip(pos_class_indices, df.T):
+        if method == 'isotonic':
+            calibrator = IsotonicRegression(out_of_bounds='clip')
+        elif method == 'sigmoid':
+            calibrator = _SigmoidCalibration()
         else:
-            self.label_encoder_.fit(self.classes)
+            raise ValueError("'method' should be one of: 'sigmoid' or "
+                             f"'isotonic'. Got {method}.")
+        calibrator.fit(this_df, Y[:, class_idx], sample_weight)
+        calibrated_classifiers.append(calibrator)
 
-        self.classes_ = self.label_encoder_.classes_
-        Y = label_binarize(y, classes=self.classes_)
+    pipeline = _CalibratedClassiferPipeline(
+        clf_fitted, calibrated_classifiers, label_encoder_
+    )
+    return pipeline
 
-        df, idx_pos_class = self._preproc(X)
-        self.calibrators_ = []
 
-        for k, this_df in zip(idx_pos_class, df.T):
-            if self.method == 'isotonic':
-                calibrator = IsotonicRegression(out_of_bounds='clip')
-            elif self.method == 'sigmoid':
-                calibrator = _SigmoidCalibration()
-            else:
-                raise ValueError('method should be "sigmoid" or '
-                                 '"isotonic". Got %s.' % self.method)
-            calibrator.fit(this_df, Y[:, k], sample_weight)
-            self.calibrators_.append(calibrator)
+class _CalibratedClassiferPipeline:
+    """Pipeline-like chaining a fitted classifier and its fitted calibrators.
 
-        return self
+    Parameters
+    ----------
+    clf_fitted : Estimator instance
+        Fitted classifier.
+
+    calibrators_fitted : List of fitted estimator instances
+        List of fitted calibrators (either 'IsotonicRegression' or
+        '_SigmoidCalibration'). The number of calibrators equals the number of
+        classes. However, if there are 2 classes, the list contains only one
+        fitted calibrator.
+    """
+    def __init__(self, clf_fitted, calibrators_fitted, label_encoder_):
```

Comment:
I think I would prefer to use the names:

- `clf_fitted` => `base_estimator`
- `calibrators_fitted` = > `calibrators`
- `label_encoder_` => `label_encoder`

and just mention in the docstring that all of them are expected to be fitted.

Also we could add a property named `calibrators_` that would just return `calibrators` and `label_encoder_` pointing to `label_encoder` to ensure backward compat with the previous attribute names, maybe with a deprecation warning.

While this is a private class, calibrated classifier instances are stored in a public attribute of the public `CalibratedClassifierCV` class, therefore could be considered semi-public API, at least from a duck-typing point of view.

Reply:
Thanks @NicolasHug reminded me of keeping the attributes constant and it's updated #17856
I've also removed the attribute `label_encoder`.

> Also we could add a property named calibrators_ that would just return calibrators

This may be tricky/confusing because for multiclass, there is more than one calibrator per classifier. e.g., for 3 classes each pair would be (1 classifier, 3 calibrators). I guess we could return a list of tuples, each tuple containing the calibrators for a 'pair' ?

Reply:
Actually I realised that `_CalibratedClassifier` has a `calibrators` attribute, which can be accessed via `calibrated_classifiers_.calibrators`